examples/pytorch/adaptive_sampling/adaptive_sampling.py

```diff
@@ -0,0 +1,448 @@
+        self.norm_adj = normalize_adj(trainG.adjacency_matrix_scipy())
+
+    def forward(self, nf, h, is_test=False):
+        for i, layer in enumerate(self.layers):
+            if i == len(self.layers) - 1 and not is_test:
+                h, loss = layer(nf, h, self.node_feature, i, self.sample_weights, norm_adj=self.norm_adj, var_loss=True,
+                                is_test=is_test)
+            else:
+                h = layer(nf, h, self.node_feature, i, self.sample_weights, is_test=is_test)
+        if is_test:
+            return h
+        return h, loss
+
+
+def main(args):
+    config.update(args)
+    config['layer_sizes'] = [int(config['node_per_layer']) for _ in range(2)]
+    # Create a sampler for training nodes and testing nodes
+    train_sampler = NodeSampler(graph=trainG, seeds=train_nodes, batch_size=config['batch_size'])
+    test_sampler = NodeSampler(graph=allG, seeds=test_nodes, batch_size=config['test_batch_size'])
+    ##Generator for training
+    train_generator = AdaptGenerator(graph=trainG, node_feature=all_h, layer_nodes=config['layer_sizes'],
+                                     sampler=train_sampler,
+                                     num_blocks=len(config['layer_sizes']), coalesce=True)
+    # Generator for testing
+    test_sample_generator = AdaptGenerator(graph=allG, node_feature=all_h, sampler=test_sampler,
+                                           num_blocks=len(config['test_layer_sizes']),
+                                           sampler_weights=train_generator.sample_weights,
+                                           layer_nodes=config['test_layer_sizes'],
+                                           coalesce=True)
+    model = AdaptGraphSAGENet(train_generator.sample_weights, all_h, config['hidden_size'])
+    params = list(model.parameters())
+    params.append(train_generator.sample_weights)
+    opt = torch.optim.Adam(params=params, lr=config['lr'])
+    # model.train()
+    lamb, weight_decay = config['lamb'], config['weight_decay']
+    for epoch in range(config['n_epoch']):
+        train_accs = []
+        for nf, sample_indices in train_generator:
+            seed_map = nf.seed_map
+            train_y_hat, varloss = model(nf, h[nf.layer_mappings[0]])
+            train_y_hat = train_y_hat[seed_map]
+            y_train_batch = y_train[sample_indices]
+            y_pred = torch.argmax(train_y_hat, dim=1)
+            train_acc = torch.sum(torch.eq(y_pred, y_train_batch)).item() / config['batch_size']
+            train_accs.append(train_acc)
+            loss = F.cross_entropy(train_y_hat.squeeze(), y_train_batch)
+            l2_loss = torch.norm(params[0])
+            total_loss = varloss * lamb + loss + l2_loss * weight_decay
+            opt.zero_grad()
+            total_loss.backward()
+            opt.step()
+            # print(train_sampler.sample_weight)
+        test_accs = []
+        for test_nf, test_sample_indices in test_sample_generator:
```

Comment:
call `model.eval` and `torch.no_grad` at evaluation stage. And `model.train` at training stage.